[Backported to Debian's 2.6.32 by dann frazier <dannf@debian.org>

commit 7dcd2499deab8f10011713c40bc2f309c9b65077
Author: Chris Wilson <chris@chris-wilson.co.uk>
Date:   Sun Sep 26 20:21:44 2010 +0100

    drm/i915: Rephrase pwrite bounds checking to avoid any potential overflow
    
    ... and do the same for pread.
    
    Signed-off-by: Chris Wilson <chris@chris-wilson.co.uk>
    Cc: stable@kernel.org

--- linux-source-2.6.32.orig/drivers/gpu/drm/i915/i915_gem.c	2010-10-11 13:35:02.000000000 -0600
+++ linux-source-2.6.32/drivers/gpu/drm/i915/i915_gem.c	2010-10-12 23:08:06.000000000 -0600
@@ -482,12 +482,8 @@ i915_gem_pread_ioctl(struct drm_device *
 		return -EBADF;
 	obj_priv = obj->driver_private;
 
-	/* Bounds check source.
-	 *
-	 * XXX: This could use review for overflow issues...
-	 */
-	if (args->offset > obj->size || args->size > obj->size ||
-	    args->offset + args->size > obj->size) {
+	/* Bounds check source.  */
+	if (args->offset > obj->size || args->size > obj->size - args->offset) {
 		ret = -EINVAL;
 		goto err;
 	}
@@ -960,12 +956,8 @@ i915_gem_pwrite_ioctl(struct drm_device
 		return -EBADF;
 	obj_priv = obj->driver_private;
 
-	/* Bounds check destination.
-	 *
-	 * XXX: This could use review for overflow issues...
-	 */
-	if (args->offset > obj->size || args->size > obj->size ||
-	    args->offset + args->size > obj->size) {
+	/* Bounds check destination. */
+	if (args->offset > obj->size || args->size > obj->size - args->offset) {
 		ret = -EINVAL;
 		goto err;
 	}
